client/service/contract_execution.go

```diff
@@ -0,0 +1,158 @@
+package service
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+
+	"github.com/google/uuid"
+	clientError "github.com/scalar-labs/scalardl-go-client-sdk/v3/client/error"
+	"github.com/scalar-labs/scalardl-go-client-sdk/v3/crypto"
+	"github.com/scalar-labs/scalardl-go-client-sdk/v3/json"
+	"github.com/scalar-labs/scalardl-go-client-sdk/v3/ledger/asset"
+	"github.com/scalar-labs/scalardl-go-client-sdk/v3/ledger/model"
+	"github.com/scalar-labs/scalardl-go-client-sdk/v3/ledger/statuscode"
+	"github.com/scalar-labs/scalardl-go-client-sdk/v3/rpc"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+// ExecuteContract executes a registered contract.
+func (s ClientService) ExecuteContract(
+	id string,
+	argument json.Object,
+	functionArgument json.Object,
+) (result model.ContractExecutionResult, err error) {
+	if s.clientConfig.ClientMode != "CLIENT" {
+		return result, clientError.NewClientError(statuscode.InvalidRequest, "wrong mode specified")
+	}
+
+	if id == "" {
+		return result, fmt.Errorf("id cannot be empty")
+	}
+
+	if argument == nil {
+		return result, fmt.Errorf("argument cannot be nil")
+	}
+
+	if _, ok := argument["nonce"]; !ok {
+		argument["nonce"] = uuid.NewString()
+	}
+
+	if argument["nonce"] == "" {
+		argument["nonce"] = uuid.NewString()
+	}
+
+	var (
+		signer  crypto.Signer
+		request = &rpc.ContractExecutionRequest{
+			ContractId:       id,
+			CertHolderId:     s.clientConfig.CertHolderID,
+			CertVersion:      uint32(s.clientConfig.CertVersion),
+			ContractArgument: argument.String(),
+		}
+	)
+
+	if functionArgument != nil {
+		request.FunctionArgument = functionArgument.String()
+	}
+
+	if signer, err = crypto.NewEcdsaSha256Signer([]byte(s.clientConfig.PrivateKey)); err != nil {
+		return
+	}
+
+	if err = request.SignWith(signer); err != nil {
+		return
+	}
+
+	var (
+		auditor             rpc.AuditorClient
+		ledger              = rpc.NewLedgerClient(s.ledgerConnection)
+		trailer             metadata.MD
+		responseFromLedger  *rpc.ContractExecutionResponse
+		responseFromAuditor *rpc.ContractExecutionResponse
+	)
+
+	// send the ordering request to Auditor first to get Auditor's signature.
+	if s.clientConfig.IsAuditorEnabled {
+		auditor = rpc.NewAuditorClient(s.auditorConnection)
+		trailer = metadata.MD{}
+
+		var ordered *rpc.ExecutionOrderingResponse
+		if ordered, err = auditor.OrderExecution(context.Background(), request, grpc.Trailer(&trailer)); err != nil {
+			if trailer.Len() > 0 {
+				err = getClientErrorFromTrailer(trailer)
+			}
+
+			return
+		}
+
+		request.AuditorSignature = ordered.GetSignature()
+	}
+
+	trailer = metadata.MD{}
+	if responseFromLedger, err = ledger.ExecuteContract(context.Background(), request, grpc.Trailer(&trailer)); err != nil {
+		if trailer.Len() > 0 {
+			err = getClientErrorFromTrailer(trailer)
+		}
+
+		return
+	}
+
+	// send the execution validation reqeust to Auditor.k
```

Comment:
Auditor.k? just a typo?

Reply:
Thank you.
I think it's because of the Vim binding.
(maybe somehow I wanna move the cursor up)